{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This package, hgvs, is an easy-to-use Python library for parsing, representing, formatting, and mapping variants between genome, transcript, and protein sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import hgvs.parser\n",
    "import hgvs.dataproviders.uta\n",
    "# Validating variants: composed of two classes: hgvs.validator.IntrinsicValidator and hgvs.validator.ExtrinsicValidator\n",
    "# Intrinsic validation evaluates a given variant for internal consistency, such as requiring that insertions specify adjacent positions. \n",
    "# Extrinsic validation evaluates a variant using external data, such as ensuring that the reference nucleotide in the variant matches that implied by the reference sequence and position.\n",
    "import hgvs.validator\n",
    "# Normalization is always 3â€™ with respect to the reference sequence. \n",
    "import hgvs.normalizer\n",
    "from hgvs.exceptions import HGVSError\n",
    "\n",
    "\n",
    "hp = hgvs.parser.Parser()\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "hn = hgvs.normalizer.Normalizer(hdp)\n",
    "vr = hgvs.validator.Validator(hdp=hdp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_hgvs_variants(expression):\n",
    "    try:\n",
    "        parsed_variant = hp.parse_hgvs_variant(expression)\n",
    "        vr.validate(parsed_variant)\n",
    "        return expression\n",
    "    except HGVSError as e:\n",
    "        return e\n",
    "validate_hgvs_variants('NC_000001.10:g.>A')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [],
   "source": [
    "excel_file = '../data/LabHgvsExpression.xlsx'\n",
    "\n",
    "with pd.ExcelFile(excel_file) as hgvs_path:\n",
    "    input_data = pd.read_excel(hgvs_path)\n",
    "\n",
    "input_data['oringal_hgvs_expression_example'] = input_data['oringal_hgvs_expression_example'].str.strip()\n",
    "input_data['edited_hgvs_expression_expample'] = input_data['edited_hgvs_expression_expample'].str.strip()\n",
    "\n",
    "data = pd.DataFrame(input_data)\n",
    "\n",
    "\n",
    "\n",
    "#NOTE: Boolean Values where obtained from: https://mutalyzer.nl/\n",
    "# Description: The Normalizer takes a variant description as input and checks whether it is correct.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oringal_hgvs_expression_example</th>\n",
       "      <th>edited_hgvs_expression_expample</th>\n",
       "      <th>mutalyzer_original_hgvs_checker</th>\n",
       "      <th>mutalyzer_edited_hgvs_checker</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>variant in the TTN(NM_001267550.2) gene</td>\n",
       "      <td>TTN(NM_001267550.2)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A p.Ser26669Asn</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001267550.2 p.Ser26669Asn</td>\n",
       "      <td>NM_001267550.2:p.Ser26669Asn</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C (p.Cys1913Arg)</td>\n",
       "      <td>NM_000059.4:c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c.5737T&gt;C (p.C1913R) in BRCA2</td>\n",
       "      <td>c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BRCA2:c.5737T&gt;C p.C1913R</td>\n",
       "      <td>BRCA2:p.C1913R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BRCA2 c.5737T&gt;C</td>\n",
       "      <td>BRCA2:c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BRCA2:C1913R</td>\n",
       "      <td>BRCA2:C1913R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C (p.Glu74Lys)</td>\n",
       "      <td>NM_000371.4:c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.E74K (Glu74Lys) c.220G&gt;C (g.29175102)</td>\n",
       "      <td>c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chr18:g.29175102</td>\n",
       "      <td>NC_000008.11:g.29175102</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTR E74K</td>\n",
       "      <td>TTR:E74K</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTR c.220G&gt;C</td>\n",
       "      <td>TTR:c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A (p.Val142Ile)</td>\n",
       "      <td>NM_000371.4:c.424G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p.V142I g.29178618</td>\n",
       "      <td>p.V142I</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c.424G&gt;A p.Val142Ile</td>\n",
       "      <td>c.424G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TTR g.29178618</td>\n",
       "      <td>TTR:g.29178618</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GRCh37 Chr18:g.29178618</td>\n",
       "      <td>NC_000018.9:g.29178618</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   oringal_hgvs_expression_example  \\\n",
       "0        NM_001267550.2:c.80006G>A (p.Ser26669Asn)   \n",
       "1                       c.80006G>A (p.Ser26669Asn)   \n",
       "2          variant in the TTN(NM_001267550.2) gene   \n",
       "3   NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn)   \n",
       "4          NM_001267550.2:c.80006G>A p.Ser26669Asn   \n",
       "5                     NM_001267550.2 p.Ser26669Asn   \n",
       "6             NM_000059.4:c.5737T>C (p.Cys1913Arg)   \n",
       "7                    c.5737T>C (p.C1913R) in BRCA2   \n",
       "8                         BRCA2:c.5737T>C p.C1913R   \n",
       "9                                  BRCA2 c.5737T>C   \n",
       "10                                    BRCA2:C1913R   \n",
       "11               NM_000371.4:c.220G>C (p.Glu74Lys)   \n",
       "12         p.E74K (Glu74Lys) c.220G>C (g.29175102)   \n",
       "13                                Chr18:g.29175102   \n",
       "14                                        TTR E74K   \n",
       "15                                    TTR c.220G>C   \n",
       "16              NM_000371.4:c.424G>A (p.Val142Ile)   \n",
       "17                              p.V142I g.29178618   \n",
       "18                            c.424G>A p.Val142Ile   \n",
       "19                                  TTR g.29178618   \n",
       "20                         GRCh37 Chr18:g.29178618   \n",
       "\n",
       "   edited_hgvs_expression_expample  mutalyzer_original_hgvs_checker  \\\n",
       "0        NM_001267550.2:c.80006G>A                            False   \n",
       "1                       c.80006G>A                            False   \n",
       "2              TTN(NM_001267550.2)                            False   \n",
       "3   NM_001267550.2(TTN):c.80006G>A                            False   \n",
       "4        NM_001267550.2:c.80006G>A                            False   \n",
       "5     NM_001267550.2:p.Ser26669Asn                            False   \n",
       "6            NM_000059.4:c.5737T>C                            False   \n",
       "7                        c.5737T>C                            False   \n",
       "8                   BRCA2:p.C1913R                            False   \n",
       "9                  BRCA2:c.5737T>C                            False   \n",
       "10                    BRCA2:C1913R                            False   \n",
       "11            NM_000371.4:c.220G>C                            False   \n",
       "12                        c.220G>C                            False   \n",
       "13         NC_000008.11:g.29175102                            False   \n",
       "14                        TTR:E74K                            False   \n",
       "15                    TTR:c.220G>C                            False   \n",
       "16            NM_000371.4:c.424G>A                            False   \n",
       "17                         p.V142I                            False   \n",
       "18                        c.424G>A                            False   \n",
       "19                  TTR:g.29178618                            False   \n",
       "20          NC_000018.9:g.29178618                             True   \n",
       "\n",
       "    mutalyzer_edited_hgvs_checker  \n",
       "0                            True  \n",
       "1                           False  \n",
       "2                           False  \n",
       "3                            True  \n",
       "4                            True  \n",
       "5                            True  \n",
       "6                            True  \n",
       "7                           False  \n",
       "8                           False  \n",
       "9                           False  \n",
       "10                          False  \n",
       "11                           True  \n",
       "12                          False  \n",
       "13                           True  \n",
       "14                          False  \n",
       "15                          False  \n",
       "16                           True  \n",
       "17                          False  \n",
       "18                          False  \n",
       "19                          False  \n",
       "20                           True  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_expression = data['oringal_hgvs_expression_example']\n",
    "edited_expression = data['edited_hgvs_expression_expample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_hgvs_variants(hgvs_list):\n",
    "    # boolean_checker = []\n",
    "    error_messages = []\n",
    "\n",
    "    for hgvs in hgvs_list:\n",
    "        try:\n",
    "            parsed_variant = hp.parse_hgvs_variant(hgvs)\n",
    "            vr.validate(parsed_variant)\n",
    "            # boolean_checker.append()\n",
    "            error_messages.append('Passed')\n",
    "        except HGVSError as e:\n",
    "            # boolean_checker.append(False)\n",
    "            error_messages.append(e) \n",
    "    \n",
    "    return boolean_checker, error_messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_boolean_checker, original_error_messages = validate_hgvs_variants(original_expression)\n",
    "edited_boolean_checker, edited_error_messages = validate_hgvs_variants(edited_expression)\n",
    "\n",
    "hgvs_results = {\n",
    "    # 'biocommons_hgvs_original_hgvs_checker':original_boolean_checker,\n",
    "    'biocommons_hgvs_original_hgvs_error_messages':original_error_messages,\n",
    "    # 'biocommons_hgvs_edited_hgvs_checker':edited_boolean_checker,\n",
    "    'biocommons_hgvs_edited_hgvs_error_messages':edited_error_messages\n",
    "}\n",
    "\n",
    "biocommons_hgvs_test = pd.DataFrame(hgvs_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>biocommons_hgvs_original_hgvs_error_messages</th>\n",
       "      <th>biocommons_hgvs_edited_hgvs_error_messages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A (p.Ser26669Asn): cha...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c.80006G&gt;A (p.Ser26669Asn): char 1: expected a...</td>\n",
       "      <td>c.80006G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>variant in the TTN(NM_001267550.2) gene: char ...</td>\n",
       "      <td>TTN(NM_001267550.2): char 6: expected one of '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A (p.Ser26669Asn)...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A p.Ser26669Asn: char ...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001267550.2 p.Ser26669Asn: char 14: expecte...</td>\n",
       "      <td>Accession (NM_001267550.2) is not compatible w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C (p.Cys1913Arg): char 21:...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000059.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c.5737T&gt;C (p.C1913R) in BRCA2: char 1: expecte...</td>\n",
       "      <td>c.5737T&gt;C: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BRCA2:c.5737T&gt;C p.C1913R: char 15: expected EOF</td>\n",
       "      <td>Accession (BRCA2) is not known to be compatibl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BRCA2 c.5737T&gt;C: char 5: expected one of '(', ...</td>\n",
       "      <td>Accession (BRCA2) is not known to be compatibl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BRCA2:C1913R: char 6: expected one of 'c', 'g'...</td>\n",
       "      <td>BRCA2:C1913R: char 6: expected one of 'c', 'g'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C (p.Glu74Lys): char 20: ex...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.E74K (Glu74Lys) c.220G&gt;C (g.29175102): char ...</td>\n",
       "      <td>c.220G&gt;C: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chr18:g.29175102: char 17: end of input</td>\n",
       "      <td>NC_000008.11:g.29175102: char 24: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTR E74K: char 3: expected one of '(', '.', ':...</td>\n",
       "      <td>TTR:E74K: char 4: expected one of 'c', 'g', 'm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTR c.220G&gt;C: char 3: expected one of '(', '.'...</td>\n",
       "      <td>Accession (TTR) is not known to be compatible ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A (p.Val142Ile): char 20: e...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p.V142I g.29178618: char 1: expected a letter ...</td>\n",
       "      <td>p.V142I: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c.424G&gt;A p.Val142Ile: char 1: expected a lette...</td>\n",
       "      <td>c.424G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TTR g.29178618: char 3: expected one of '(', '...</td>\n",
       "      <td>TTR:g.29178618: char 15: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GRCh37 Chr18:g.29178618: char 6: expected one ...</td>\n",
       "      <td>NC_000018.9:g.29178618: char 23: end of input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         biocommons_hgvs_original_hgvs_error_messages  \\\n",
       "0   NM_001267550.2:c.80006G>A (p.Ser26669Asn): cha...   \n",
       "1   c.80006G>A (p.Ser26669Asn): char 1: expected a...   \n",
       "2   variant in the TTN(NM_001267550.2) gene: char ...   \n",
       "3   NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn)...   \n",
       "4   NM_001267550.2:c.80006G>A p.Ser26669Asn: char ...   \n",
       "5   NM_001267550.2 p.Ser26669Asn: char 14: expecte...   \n",
       "6   NM_000059.4:c.5737T>C (p.Cys1913Arg): char 21:...   \n",
       "7   c.5737T>C (p.C1913R) in BRCA2: char 1: expecte...   \n",
       "8     BRCA2:c.5737T>C p.C1913R: char 15: expected EOF   \n",
       "9   BRCA2 c.5737T>C: char 5: expected one of '(', ...   \n",
       "10  BRCA2:C1913R: char 6: expected one of 'c', 'g'...   \n",
       "11  NM_000371.4:c.220G>C (p.Glu74Lys): char 20: ex...   \n",
       "12  p.E74K (Glu74Lys) c.220G>C (g.29175102): char ...   \n",
       "13            Chr18:g.29175102: char 17: end of input   \n",
       "14  TTR E74K: char 3: expected one of '(', '.', ':...   \n",
       "15  TTR c.220G>C: char 3: expected one of '(', '.'...   \n",
       "16  NM_000371.4:c.424G>A (p.Val142Ile): char 20: e...   \n",
       "17  p.V142I g.29178618: char 1: expected a letter ...   \n",
       "18  c.424G>A p.Val142Ile: char 1: expected a lette...   \n",
       "19  TTR g.29178618: char 3: expected one of '(', '...   \n",
       "20  GRCh37 Chr18:g.29178618: char 6: expected one ...   \n",
       "\n",
       "           biocommons_hgvs_edited_hgvs_error_messages  \n",
       "0                                              Passed  \n",
       "1      c.80006G>A: char 1: expected a letter or digit  \n",
       "2   TTN(NM_001267550.2): char 6: expected one of '...  \n",
       "3                                              Passed  \n",
       "4                                              Passed  \n",
       "5   Accession (NM_001267550.2) is not compatible w...  \n",
       "6    No transcript definition for (tx_ac=NM_000059.4)  \n",
       "7       c.5737T>C: char 1: expected a letter or digit  \n",
       "8   Accession (BRCA2) is not known to be compatibl...  \n",
       "9   Accession (BRCA2) is not known to be compatibl...  \n",
       "10  BRCA2:C1913R: char 6: expected one of 'c', 'g'...  \n",
       "11   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "12       c.220G>C: char 1: expected a letter or digit  \n",
       "13     NC_000008.11:g.29175102: char 24: end of input  \n",
       "14  TTR:E74K: char 4: expected one of 'c', 'g', 'm...  \n",
       "15  Accession (TTR) is not known to be compatible ...  \n",
       "16   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "17        p.V142I: char 1: expected a letter or digit  \n",
       "18       c.424G>A: char 1: expected a letter or digit  \n",
       "19              TTR:g.29178618: char 15: end of input  \n",
       "20      NC_000018.9:g.29178618: char 23: end of input  "
      ]
     },
     "execution_count": 176,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biocommons_hgvs_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>oringal_hgvs_expression_example</th>\n",
       "      <th>edited_hgvs_expression_expample</th>\n",
       "      <th>mutalyzer_original_hgvs_checker</th>\n",
       "      <th>mutalyzer_edited_hgvs_checker</th>\n",
       "      <th>biocommons_hgvs_original_hgvs_error_messages</th>\n",
       "      <th>biocommons_hgvs_edited_hgvs_error_messages</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A (p.Ser26669Asn): cha...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>c.80006G&gt;A (p.Ser26669Asn): char 1: expected a...</td>\n",
       "      <td>c.80006G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>variant in the TTN(NM_001267550.2) gene</td>\n",
       "      <td>TTN(NM_001267550.2)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>variant in the TTN(NM_001267550.2) gene: char ...</td>\n",
       "      <td>TTN(NM_001267550.2): char 6: expected one of '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A (p.Ser26669Asn)...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A p.Ser26669Asn</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A p.Ser26669Asn: char ...</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001267550.2 p.Ser26669Asn</td>\n",
       "      <td>NM_001267550.2:p.Ser26669Asn</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_001267550.2 p.Ser26669Asn: char 14: expecte...</td>\n",
       "      <td>Accession (NM_001267550.2) is not compatible w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C (p.Cys1913Arg)</td>\n",
       "      <td>NM_000059.4:c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_000059.4:c.5737T&gt;C (p.Cys1913Arg): char 21:...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000059.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c.5737T&gt;C (p.C1913R) in BRCA2</td>\n",
       "      <td>c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>c.5737T&gt;C (p.C1913R) in BRCA2: char 1: expecte...</td>\n",
       "      <td>c.5737T&gt;C: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BRCA2:c.5737T&gt;C p.C1913R</td>\n",
       "      <td>BRCA2:p.C1913R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BRCA2:c.5737T&gt;C p.C1913R: char 15: expected EOF</td>\n",
       "      <td>Accession (BRCA2) is not known to be compatibl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BRCA2 c.5737T&gt;C</td>\n",
       "      <td>BRCA2:c.5737T&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BRCA2 c.5737T&gt;C: char 5: expected one of '(', ...</td>\n",
       "      <td>Accession (BRCA2) is not known to be compatibl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BRCA2:C1913R</td>\n",
       "      <td>BRCA2:C1913R</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>BRCA2:C1913R: char 6: expected one of 'c', 'g'...</td>\n",
       "      <td>BRCA2:C1913R: char 6: expected one of 'c', 'g'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C (p.Glu74Lys)</td>\n",
       "      <td>NM_000371.4:c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_000371.4:c.220G&gt;C (p.Glu74Lys): char 20: ex...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.E74K (Glu74Lys) c.220G&gt;C (g.29175102)</td>\n",
       "      <td>c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>p.E74K (Glu74Lys) c.220G&gt;C (g.29175102): char ...</td>\n",
       "      <td>c.220G&gt;C: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chr18:g.29175102</td>\n",
       "      <td>NC_000008.11:g.29175102</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>Chr18:g.29175102: char 17: end of input</td>\n",
       "      <td>NC_000008.11:g.29175102: char 24: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTR E74K</td>\n",
       "      <td>TTR:E74K</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TTR E74K: char 3: expected one of '(', '.', ':...</td>\n",
       "      <td>TTR:E74K: char 4: expected one of 'c', 'g', 'm...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTR c.220G&gt;C</td>\n",
       "      <td>TTR:c.220G&gt;C</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TTR c.220G&gt;C: char 3: expected one of '(', '.'...</td>\n",
       "      <td>Accession (TTR) is not known to be compatible ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A (p.Val142Ile)</td>\n",
       "      <td>NM_000371.4:c.424G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>NM_000371.4:c.424G&gt;A (p.Val142Ile): char 20: e...</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p.V142I g.29178618</td>\n",
       "      <td>p.V142I</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>p.V142I g.29178618: char 1: expected a letter ...</td>\n",
       "      <td>p.V142I: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c.424G&gt;A p.Val142Ile</td>\n",
       "      <td>c.424G&gt;A</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>c.424G&gt;A p.Val142Ile: char 1: expected a lette...</td>\n",
       "      <td>c.424G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TTR g.29178618</td>\n",
       "      <td>TTR:g.29178618</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>TTR g.29178618: char 3: expected one of '(', '...</td>\n",
       "      <td>TTR:g.29178618: char 15: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GRCh37 Chr18:g.29178618</td>\n",
       "      <td>NC_000018.9:g.29178618</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>GRCh37 Chr18:g.29178618: char 6: expected one ...</td>\n",
       "      <td>NC_000018.9:g.29178618: char 23: end of input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   oringal_hgvs_expression_example  \\\n",
       "0        NM_001267550.2:c.80006G>A (p.Ser26669Asn)   \n",
       "1                       c.80006G>A (p.Ser26669Asn)   \n",
       "2          variant in the TTN(NM_001267550.2) gene   \n",
       "3   NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn)   \n",
       "4          NM_001267550.2:c.80006G>A p.Ser26669Asn   \n",
       "5                     NM_001267550.2 p.Ser26669Asn   \n",
       "6             NM_000059.4:c.5737T>C (p.Cys1913Arg)   \n",
       "7                    c.5737T>C (p.C1913R) in BRCA2   \n",
       "8                         BRCA2:c.5737T>C p.C1913R   \n",
       "9                                  BRCA2 c.5737T>C   \n",
       "10                                    BRCA2:C1913R   \n",
       "11               NM_000371.4:c.220G>C (p.Glu74Lys)   \n",
       "12         p.E74K (Glu74Lys) c.220G>C (g.29175102)   \n",
       "13                                Chr18:g.29175102   \n",
       "14                                        TTR E74K   \n",
       "15                                    TTR c.220G>C   \n",
       "16              NM_000371.4:c.424G>A (p.Val142Ile)   \n",
       "17                              p.V142I g.29178618   \n",
       "18                            c.424G>A p.Val142Ile   \n",
       "19                                  TTR g.29178618   \n",
       "20                         GRCh37 Chr18:g.29178618   \n",
       "\n",
       "   edited_hgvs_expression_expample  mutalyzer_original_hgvs_checker  \\\n",
       "0        NM_001267550.2:c.80006G>A                            False   \n",
       "1                       c.80006G>A                            False   \n",
       "2              TTN(NM_001267550.2)                            False   \n",
       "3   NM_001267550.2(TTN):c.80006G>A                            False   \n",
       "4        NM_001267550.2:c.80006G>A                            False   \n",
       "5     NM_001267550.2:p.Ser26669Asn                            False   \n",
       "6            NM_000059.4:c.5737T>C                            False   \n",
       "7                        c.5737T>C                            False   \n",
       "8                   BRCA2:p.C1913R                            False   \n",
       "9                  BRCA2:c.5737T>C                            False   \n",
       "10                    BRCA2:C1913R                            False   \n",
       "11            NM_000371.4:c.220G>C                            False   \n",
       "12                        c.220G>C                            False   \n",
       "13         NC_000008.11:g.29175102                            False   \n",
       "14                        TTR:E74K                            False   \n",
       "15                    TTR:c.220G>C                            False   \n",
       "16            NM_000371.4:c.424G>A                            False   \n",
       "17                         p.V142I                            False   \n",
       "18                        c.424G>A                            False   \n",
       "19                  TTR:g.29178618                            False   \n",
       "20          NC_000018.9:g.29178618                             True   \n",
       "\n",
       "    mutalyzer_edited_hgvs_checker  \\\n",
       "0                            True   \n",
       "1                           False   \n",
       "2                           False   \n",
       "3                            True   \n",
       "4                            True   \n",
       "5                            True   \n",
       "6                            True   \n",
       "7                           False   \n",
       "8                           False   \n",
       "9                           False   \n",
       "10                          False   \n",
       "11                           True   \n",
       "12                          False   \n",
       "13                           True   \n",
       "14                          False   \n",
       "15                          False   \n",
       "16                           True   \n",
       "17                          False   \n",
       "18                          False   \n",
       "19                          False   \n",
       "20                           True   \n",
       "\n",
       "         biocommons_hgvs_original_hgvs_error_messages  \\\n",
       "0   NM_001267550.2:c.80006G>A (p.Ser26669Asn): cha...   \n",
       "1   c.80006G>A (p.Ser26669Asn): char 1: expected a...   \n",
       "2   variant in the TTN(NM_001267550.2) gene: char ...   \n",
       "3   NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn)...   \n",
       "4   NM_001267550.2:c.80006G>A p.Ser26669Asn: char ...   \n",
       "5   NM_001267550.2 p.Ser26669Asn: char 14: expecte...   \n",
       "6   NM_000059.4:c.5737T>C (p.Cys1913Arg): char 21:...   \n",
       "7   c.5737T>C (p.C1913R) in BRCA2: char 1: expecte...   \n",
       "8     BRCA2:c.5737T>C p.C1913R: char 15: expected EOF   \n",
       "9   BRCA2 c.5737T>C: char 5: expected one of '(', ...   \n",
       "10  BRCA2:C1913R: char 6: expected one of 'c', 'g'...   \n",
       "11  NM_000371.4:c.220G>C (p.Glu74Lys): char 20: ex...   \n",
       "12  p.E74K (Glu74Lys) c.220G>C (g.29175102): char ...   \n",
       "13            Chr18:g.29175102: char 17: end of input   \n",
       "14  TTR E74K: char 3: expected one of '(', '.', ':...   \n",
       "15  TTR c.220G>C: char 3: expected one of '(', '.'...   \n",
       "16  NM_000371.4:c.424G>A (p.Val142Ile): char 20: e...   \n",
       "17  p.V142I g.29178618: char 1: expected a letter ...   \n",
       "18  c.424G>A p.Val142Ile: char 1: expected a lette...   \n",
       "19  TTR g.29178618: char 3: expected one of '(', '...   \n",
       "20  GRCh37 Chr18:g.29178618: char 6: expected one ...   \n",
       "\n",
       "           biocommons_hgvs_edited_hgvs_error_messages  \n",
       "0                                              Passed  \n",
       "1      c.80006G>A: char 1: expected a letter or digit  \n",
       "2   TTN(NM_001267550.2): char 6: expected one of '...  \n",
       "3                                              Passed  \n",
       "4                                              Passed  \n",
       "5   Accession (NM_001267550.2) is not compatible w...  \n",
       "6    No transcript definition for (tx_ac=NM_000059.4)  \n",
       "7       c.5737T>C: char 1: expected a letter or digit  \n",
       "8   Accession (BRCA2) is not known to be compatibl...  \n",
       "9   Accession (BRCA2) is not known to be compatibl...  \n",
       "10  BRCA2:C1913R: char 6: expected one of 'c', 'g'...  \n",
       "11   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "12       c.220G>C: char 1: expected a letter or digit  \n",
       "13     NC_000008.11:g.29175102: char 24: end of input  \n",
       "14  TTR:E74K: char 4: expected one of 'c', 'g', 'm...  \n",
       "15  Accession (TTR) is not known to be compatible ...  \n",
       "16   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "17        p.V142I: char 1: expected a letter or digit  \n",
       "18       c.424G>A: char 1: expected a letter or digit  \n",
       "19              TTR:g.29178618: char 15: end of input  \n",
       "20      NC_000018.9:g.29178618: char 23: end of input  "
      ]
     },
     "execution_count": 177,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.concat([data, biocommons_hgvs_test], axis=1, ignore_index=False)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv('../data/results.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No transcript definition for (tx_ac=NM_000371.4)\n"
     ]
    }
   ],
   "source": [
    "import hgvs.validator\n",
    "import hgvs.exceptions\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "try:\n",
    "    vr.validate( hp.parse_hgvs_variant('NM_000371.4:c.220G>C') )\n",
    "except hgvs.exceptions.HGVSError as e:\n",
    "    print(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hgvs_list = ['NC_000017.11:g.43091687delC', 'NC_000007.13.g.21726874G>A']\n",
    "\n",
    "try:\n",
    "    parser = hgvs.parser.Parser()\n",
    "    validator = hgvs.validator.IntrinsicValidator()\n",
    "    \n",
    "    for hgvs in hgvs_list:\n",
    "        variant = parser.parse_hgvs_variant(hgvs)\n",
    "        validator.validate(variant)\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "\n",
    "import hgvs.validator\n",
    "import hgvs.exceptions\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "try:\n",
    "    vr.validate( hp.parse_hgvs_variant('NM_001267550.2(TTN):c.80006G>A') )\n",
    "except hgvs.exceptions.HGVSError as e:\n",
    "    print(e)\n",
    "\n",
    "\n",
    "from hgvs.exceptions import HGVSError\n",
    "import hgvs.parser\n",
    "import hgvs.validator\n",
    "\n",
    "hgvs_list = ('NC_000017.11:g.43091687delC', 'NC_000007.13.g.21726874G>A')\n",
    "\n",
    "# Create a validator instance\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "# Create a parser instance\n",
    "hp = hgvs.parser.Parser()\n",
    "\n",
    "try:\n",
    "    for hgvs in hgvs_list:\n",
    "        vr.validate(hp.parse_hgvs_variant(hgvs))\n",
    "        print('expression passed: {}'.format(hgvs))\n",
    "except HGVSError as e:\n",
    "    print(e)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_dev_vrs_installation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
