{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/M278428/opt/anaconda3/envs/test_dev_vrs_installation/lib/python3.9/site-packages/python_jsonschema_objects/__init__.py:49: UserWarning: Schema version http://json-schema.org/draft-07/schema not recognized. Some keywords and features may not be supported.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from src.core_variant_translate import CVCTranslator\n",
    "from src.api.seqrepo_api import SeqRepoAPI\n",
    "cn = SeqRepoAPI(\"https://services.genomicmedlab.org/seqrepo\")\n",
    "dp = cn.dp\n",
    "varTrans = CVCTranslator()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spdi = 'NM_001267550.2:80230:G:A'\n",
    "\n",
    "print(varTrans.spdi_to_cvc(spdi))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NC_000007.14:g.55181230_55181231insGGCT\n",
      "CoreVariantClass(0-based interbase,DNA,,GGCT,55181230,55181231,None,None,None,None,None,NC_000007.14,{})\n",
      "NC_000019.10:g.44908822C>T\n",
      "CoreVariantClass(0-based interbase,DNA,C,T,44908821,44908822,None,None,None,None,None,NC_000019.10,{})\n",
      "NC_000007.14:g.55181220del\n",
      "CoreVariantClass(0-based interbase,DNA,T,,55181219,55181220,None,None,None,None,None,NC_000007.14,{})\n",
      "NC_000023.11:g.32386323delinsGA\n",
      "CoreVariantClass(0-based interbase,DNA,T,GA,32386322,32386323,None,None,None,None,None,NC_000023.11,{})\n",
      "NC_000013.11:g.32936732=\n",
      "CoreVariantClass(0-based interbase,DNA,C,C,32936731,32936732,None,None,None,None,None,NC_000013.11,{})\n",
      "NC_000013.11:g.19993838_19993839dup\n",
      "CoreVariantClass(0-based interbase,DNA,GT,GTGT,19993837,19993839,None,None,None,None,None,NC_000013.11,{})\n"
     ]
    }
   ],
   "source": [
    "hgvs = \"NC_000001.11:g.943043C>T\"\n",
    "l = [\"NC_000007.14:g.55181230_55181231insGGCT\",\"NC_000019.10:g.44908822C>T\",\"NC_000007.14:g.55181220del\",\"NC_000023.11:g.32386323delinsGA\",\"NC_000013.11:g.32936732=\",\"NC_000013.11:g.19993838_19993839dup\"]\n",
    "\n",
    "for hgvs in l:\n",
    "    print(hgvs)\n",
    "    print(varTrans.hgvs_to_cvc(hgvs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import hgvs.parser\n",
    "import hgvs.dataproviders.uta\n",
    "# Validating variants: composed of two classes: hgvs.validator.IntrinsicValidator and hgvs.validator.ExtrinsicValidator\n",
    "# Intrinsic validation evaluates a given variant for internal consistency, such as requiring that insertions specify adjacent positions. \n",
    "# Extrinsic validation evaluates a variant using external data, such as ensuring that the reference nucleotide in the variant matches that implied by the reference sequence and position.\n",
    "import hgvs.validator\n",
    "# Normalization is always 3â€™ with respect to the reference sequence. \n",
    "import hgvs.normalizer\n",
    "from hgvs.exceptions import HGVSError\n",
    "\n",
    "\n",
    "hp = hgvs.parser.Parser()\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "hn = hgvs.normalizer.Normalizer(hdp)\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "# hgvs_expr1 = \"NC_000001.11:g.943043C>T\"#\"NM_000097.7:c.814A>C\"\n",
    "\n",
    "# def val_hgvs_expression(expression):\n",
    "#     parsed_variant = hp.parse_hgvs_variant(hgvs_expr1)\n",
    "#     if vr.validate(parsed_variant):\n",
    "#         return parsed_variant \n",
    "    \n",
    "# val_hgvs_expression(hgvs_expr1)\n",
    "\n",
    "# hgvs_expr1 = \"NC_000013.11:g.32936732=\"#\"NM_000097.7:c.814A>C\"\n",
    "# parsed_variant = hp.parse_hgvs_variant(hgvs_expr1)\n",
    "# if not vr.validate(parsed_variant):\n",
    "#     raise(ValueError(\"Invalid HGVS expression: %s\" % hgvs_expr1)) \n",
    "# # parsed_variant.posedit.edit.ref\n",
    "# parsed_variant.posedit.edit.alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "l = [\"NC_000007.14:g.55181230_55181231insGGCT\",\"NC_000019.10:g.44908822C>T\",\"NC_000007.14:g.55181220del\",\"NC_000023.11:g.32386323delinsGA\",\"NC_000013.11:g.32936732=\",\"NC_000013.11:g.19993838_19993839dup\"]\n",
    "for i in l:\n",
    "    parsed_variant = hp.parse_hgvs_variant(i)\n",
    "    print(parsed_variant.posedit.edit.type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "insertion = \"NC_000007.14:g.55181230_55181231insGGCT\"\n",
    "insertion_parse = hp.parse_hgvs_variant(insertion)\n",
    "\n",
    "sequence = dp.get_sequence(insertion_parse.ac,insertion_parse.posedit.pos.start.base,insertion_parse.posedit.pos.end.base)\n",
    "print(sequence)\n",
    "print('Insertion Example')\n",
    "print(f\"ALTERNATIVE: {insertion_parse.posedit.edit.alt}, REFERENCE: {insertion_parse.posedit.edit.ref}\")\n",
    "\n",
    "\n",
    "# When the type is insertion: \n",
    "    # hgvs.posedit.edit.alt is the alternative sequence inserted at the given position\n",
    "    # hgvs.posedit.edit.ref gives back a value of NONE, Because we are just referring to just a position.\n",
    "    # so we need to be able to handle this case.\n",
    "\n",
    "# hgvs insertion convertion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"ins\":\n",
    "    # start and end are not change with insertion\n",
    "    start_pos = parsed_variant.posedit.pos.start.base\n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "    ref_seq = parsed_variant.posedit.edit.ref or \"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def testoutput(expression):\n",
    "\n",
    "#     parsed_variant = hp.parse_hgvs_variant(insertion)\n",
    "#     if parsed_variant.posedit.edit.type == \"ins\":\n",
    "#         start_pos = parsed_variant.posedit.pos.start.base\n",
    "#         end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "#         alt_seq = parsed_variant.posedit.edit.alt\n",
    "#         ref_seq = parsed_variant.posedit.edit.ref or \"\"\n",
    "#     return ref_seq\n",
    "\n",
    "# x = testoutput(insertion)\n",
    "# if x == \"\":\n",
    "#     print('yes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if insertion_parse.posedit.edit.ref == None:\n",
    "    print('this is a none value')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub = \"NC_000019.10:g.44908822C>T\"\n",
    "sub_parse = hp.parse_hgvs_variant(sub)\n",
    "print('Substitution Example')\n",
    "print(f\"ALTERNATIVE: {sub_parse.posedit.edit.alt}, REFERENCE: {sub_parse.posedit.edit.ref}\")\n",
    "\n",
    "if parsed_variant.posedit.edit.type == \"sub\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "\n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "    ref_seq = parsed_variant.posedit.edit.ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dell = \"NC_000007.14:g.55181220del\"\n",
    "dell_parse = hp.parse_hgvs_variant(dell)\n",
    "sequence_del = dp.get_sequence(dell_parse.ac,dell_parse.posedit.pos.start.base-1,dell_parse.posedit.pos.end.base)\n",
    "print(sequence_del)\n",
    "print('Deletion Example')\n",
    "print(f\"ALTERNATIVE: {dell_parse.posedit.edit.alt}, REFERENCE: {dell_parse.posedit.edit.ref}\")\n",
    "\n",
    "# reference sequence is whats being deleted. \n",
    "# the atlernative sequence is nothing\n",
    "\n",
    "if parsed_variant.posedit.edit.type == \"del\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(dell_parse.ac,dell_parse.posedit.pos.start.base-1,dell_parse.posedit.pos.end.base)\n",
    "    alt_seq = parsed_variant.posedit.edit.alt or \"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "delins = \"NC_000023.11:g.32386323delinsGA\"\n",
    "delins_parse = hp.parse_hgvs_variant(delins)\n",
    "sequence_delins_parse = dp.get_sequence(delins_parse.ac,delins_parse.posedit.pos.start.base-1,delins_parse.posedit.pos.end.base)\n",
    "print(sequence_delins_parse)\n",
    "print('DeletionInsertion Example')\n",
    "print(f\"ALTERNATIVE: {delins_parse.posedit.edit.alt}, REFERENCE: {delins_parse.posedit.edit.ref}\")\n",
    "\n",
    "if parsed_variant.posedit.edit.type == \"delins\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(dell_parse.ac,start_pos,end_pos)\n",
    "    alt_seq = parsed_variant.posedit.edit.alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "identity = \"NC_000013.11:g.32936732=\"\n",
    "identity_parse = hp.parse_hgvs_variant(identity)\n",
    "sequence_identity = dp.get_sequence(identity_parse.ac,identity_parse.posedit.pos.start.base-1,identity_parse.posedit.pos.end.base)\n",
    "\n",
    "print(sequence_identity)\n",
    "print('Identity Example')\n",
    "print(f\"ALTERNATIVE: {identity_parse.posedit.edit.alt}, REFERENCE: {identity_parse.posedit.edit.ref}\")\n",
    "\n",
    "\n",
    "# if parsed_variant.posedit.edit.type == \"identity\":\n",
    "#     # The starting position is changed to account for 0-interbase indexing\n",
    "#     start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "#     end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "#     ref_seq = parsed_variant.posedit.edit.alt\n",
    "#     alt_seq = parsed_variant.posedit.edit.alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "duplication = \"NC_000013.11:g.19993838_19993839dup\"\n",
    "duplication_parse = hp.parse_hgvs_variant(duplication)\n",
    "sequence_duplication = dp.get_sequence(duplication_parse.ac,duplication_parse.posedit.pos.start.base,duplication_parse.posedit.pos.end.base)\n",
    "\n",
    "print(sequence_duplication)\n",
    "print('duplication Example')\n",
    "# print(f\"ALTERNATIVE: {duplication_parse.posedit.edit.alt}, REFERENCE: {duplication_parse.posedit.edit.ref}\")\n",
    "\n",
    "\n",
    "ref = dp.get_sequence(duplication_parse.ac,duplication_parse.posedit.pos.start.base-1,duplication_parse.posedit.pos.end.base)\n",
    "print(ref)\n",
    "\n",
    "state = ref + ref\n",
    "print(state)\n",
    "\n",
    "if parsed_variant.posedit.edit.type == \"dup\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(duplication_parse.ac,start_pos,end_pos)\n",
    "    alt_seq = ref_seq + ref_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HGVS insertion conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"ins\":\n",
    "    # start and end are not change with insertion\n",
    "    start_pos = parsed_variant.posedit.pos.start.base\n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "    ref_seq = parsed_variant.posedit.edit.ref or \"\"\n",
    "    \n",
    "# HGVS substitution conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"sub\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "\n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "    ref_seq = parsed_variant.posedit.edit.ref\n",
    "    \n",
    "# HGVS deletion conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"del\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(dell_parse.ac,dell_parse.posedit.pos.start.base-1,dell_parse.posedit.pos.end.base)\n",
    "    alt_seq = parsed_variant.posedit.edit.alt or \"\"\n",
    "\n",
    "# HGVS deletion insertion conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"delins\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(dell_parse.ac,start_pos,end_pos)\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "\n",
    "# HGVS identity conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"identity\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = parsed_variant.posedit.edit.alt\n",
    "    alt_seq = parsed_variant.posedit.edit.alt\n",
    "    \n",
    "# HGVS duplication conversion to CVC format\n",
    "if parsed_variant.posedit.edit.type == \"dup\":\n",
    "    # The starting position is changed to account for 0-interbase indexing\n",
    "    start_pos = parsed_variant.posedit.pos.start.base - 1 \n",
    "    end_pos = parsed_variant.posedit.pos.end.base\n",
    "\n",
    "    ref_seq = dp.get_sequence(duplication_parse.ac,start_pos,end_pos)\n",
    "    alt_seq = ref_seq + ref_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "illegal target for annotation (3203764587.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Cell \u001b[0;32mIn [4], line 1\u001b[0;36m\u001b[0m\n\u001b[0;31m    \"search_term\": \"NC_000007.14:g.55181230_55181231insGGCT\",\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m illegal target for annotation\n"
     ]
    }
   ],
   "source": [
    "# \"search_term\": \"NC_000007.14:g.55181230_55181231insGGCT\",\n",
    "\n",
    "# {\n",
    "#     \"_id\": \"ga4gh:VA.JKGCs07cFu2wlDydCAe2ea06jMFXyK56\",\n",
    "#     \"type\": \"Allele\",\n",
    "#     \"location\": {\n",
    "#     \"_id\": \"ga4gh:VSL.SdvAZCNKh5kf6ClsiOOmw_88fbkFPTqG\",\n",
    "#     \"type\": \"SequenceLocation\",\n",
    "#     \"sequence_id\": \"ga4gh:SQ.F-LrLMe1SRpfUZHkQmvkVKFEGaoDeHul\",\n",
    "#     \"interval\": {\n",
    "#         \"type\": \"SequenceInterval\",\n",
    "#         \"start\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 55181230\n",
    "#         },\n",
    "#         \"end\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 55181230\n",
    "#         }\n",
    "#     }\n",
    "#     },\n",
    "#     \"state\": {\n",
    "#     \"type\": \"LiteralSequenceExpression\",\n",
    "#     \"sequence\": \"GGCT\"\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# \"search_term\": \"NC_000019.10:g.44908822C>T\"\n",
    "\n",
    "\n",
    "# {\n",
    "#     \"_id\": \"ga4gh:VA.CxiA_hvYbkD8Vqwjhx5AYuyul4mtlkpD\",\n",
    "#     \"type\": \"Allele\",\n",
    "#     \"location\": {\n",
    "#     \"_id\": \"ga4gh:VSL.QrRSuBj-VScAGV_gEdxNgsnh41jYH1Kg\",\n",
    "#     \"type\": \"SequenceLocation\",\n",
    "#     \"sequence_id\": \"ga4gh:SQ.IIB53T8CNeJJdUqzn9V_JnRtQadwWCbl\",\n",
    "#     \"interval\": {\n",
    "#         \"type\": \"SequenceInterval\",\n",
    "#         \"start\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 44908821\n",
    "#         },\n",
    "#         \"end\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 44908822\n",
    "#         }\n",
    "#     }\n",
    "#     },\n",
    "#     \"state\": {\n",
    "#     \"type\": \"LiteralSequenceExpression\",\n",
    "#     \"sequence\": \"T\"\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# \"search_term\": \"NC_000007.14:g.55181220del\"\n",
    "\n",
    "# {\n",
    "#     \"_id\": \"ga4gh:VA.h6WuolTwZJYZh86qP2a8YVA1WXpHuY_X\",\n",
    "#     \"type\": \"Allele\",\n",
    "#     \"location\": {\n",
    "#     \"_id\": \"ga4gh:VSL.kB_ok6Eka0225QwwbOKtvcYZBz7Z0mSR\",\n",
    "#     \"type\": \"SequenceLocation\",\n",
    "#     \"sequence_id\": \"ga4gh:SQ.F-LrLMe1SRpfUZHkQmvkVKFEGaoDeHul\",\n",
    "#     \"interval\": {\n",
    "#         \"type\": \"SequenceInterval\",\n",
    "#         \"start\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 55181219\n",
    "#         },\n",
    "#         \"end\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 55181220\n",
    "#         }\n",
    "#     }\n",
    "#     },\n",
    "#     \"state\": {\n",
    "#     \"type\": \"LiteralSequenceExpression\",\n",
    "#     \"sequence\": \"\"\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# \"search_term\": \"NC_000023.11:g.32386323delinsGA\",\n",
    "\n",
    "# {\n",
    "#     \"_id\": \"ga4gh:VA.HH3RHjZymrie-09X8aR2SMf1ULMlee6u\",\n",
    "#     \"type\": \"Allele\",\n",
    "#     \"location\": {\n",
    "#     \"_id\": \"ga4gh:VSL.tS45HvJapFexhxmbHe6SBn7dGuC46sni\",\n",
    "#     \"type\": \"SequenceLocation\",\n",
    "#     \"sequence_id\": \"ga4gh:SQ.w0WZEvgJF0zf_P4yyTzjjv9oW1z61HHP\",\n",
    "#     \"interval\": {\n",
    "#         \"type\": \"SequenceInterval\",\n",
    "#         \"start\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 32386322\n",
    "#         },\n",
    "#         \"end\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 32386323\n",
    "#         }\n",
    "#     }\n",
    "#     },\n",
    "#     \"state\": {\n",
    "#     \"type\": \"LiteralSequenceExpression\",\n",
    "#     \"sequence\": \"GA\"\n",
    "#     }\n",
    "# }\n",
    "\n",
    "# \"search_term\": \"NC_000013.11:g.32936732=\"\n",
    "# {\n",
    "#     \"_id\": \"ga4gh:VA.DkZLLMnwoH6zIncSRh2c05nzCNLdTqHl\",\n",
    "#     \"type\": \"Allele\",\n",
    "#     \"location\": {\n",
    "#     \"_id\": \"ga4gh:VSL.iSZclbNW8T95cXDuNvLMvm6xJd2g4pTn\",\n",
    "#     \"type\": \"SequenceLocation\",\n",
    "#     \"sequence_id\": \"ga4gh:SQ._0wi-qoDrvram155UmcSC-zA5ZK4fpLT\",\n",
    "#     \"interval\": {\n",
    "#         \"type\": \"SequenceInterval\",\n",
    "#         \"start\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 32936731\n",
    "#         },\n",
    "#         \"end\": {\n",
    "#         \"type\": \"Number\",\n",
    "#         \"value\": 32936732\n",
    "#         }\n",
    "#     }\n",
    "#     },\n",
    "#     \"state\": {\n",
    "#     \"type\": \"LiteralSequenceExpression\",\n",
    "#     \"sequence\": \"C\"\n",
    "#     }\n",
    "# }\n",
    "\n",
    "\n",
    "\"search_term\": \"NC_000013.11:g.19993838_19993839dup\",\n",
    "\n",
    "{\n",
    "    \"_id\": \"ga4gh:VA.S3eUS2hlp6q4pSv4u2CbN0OPMusMUnHZ\",\n",
    "    \"type\": \"Allele\",\n",
    "    \"location\": {\n",
    "    \"_id\": \"ga4gh:VSL.nW80UuWsc9bgMncP24FLII6qx8aouNki\",\n",
    "    \"type\": \"SequenceLocation\",\n",
    "    \"sequence_id\": \"ga4gh:SQ._0wi-qoDrvram155UmcSC-zA5ZK4fpLT\",\n",
    "    \"interval\": {\n",
    "        \"type\": \"SequenceInterval\",\n",
    "        \"start\": {\n",
    "        \"type\": \"Number\",\n",
    "        \"value\": 19993837\n",
    "        },\n",
    "        \"end\": {\n",
    "        \"type\": \"Number\",\n",
    "        \"value\": 19993839\n",
    "        }\n",
    "    }\n",
    "    },\n",
    "    \"state\": {\n",
    "    \"type\": \"LiteralSequenceExpression\",\n",
    "    \"sequence\": \"GTGT\"\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GT'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ref_sequence_id = translated_sequence_ids[0].split(\":\")[1]\n",
    "reference_allele = str(\n",
    "    dp.get_sequence(\n",
    "        'NC_000013.11',\n",
    "        19993837,\n",
    "        19993839,\n",
    "    )\n",
    ")\n",
    "reference_allele\n",
    "# TODO: look into this: https://github.com/ga4gh/vrs-python/blob/593508c6e8229336ca1f53a06f69966020cd68f7/src/ga4gh/vrs/extras/translator.py#L412\n",
    "# alternative_allele = (str(expression.state.sequence) if expression.state.sequence else None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_dev_vrs_installation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
