{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This package, hgvs, is an easy-to-use Python library for parsing, representing, formatting, and mapping variants between genome, transcript, and protein sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import hgvs.parser\n",
    "import hgvs.dataproviders.uta\n",
    "# Validating variants: composed of two classes: hgvs.validator.IntrinsicValidator and hgvs.validator.ExtrinsicValidator\n",
    "# Intrinsic validation evaluates a given variant for internal consistency, such as requiring that insertions specify adjacent positions. \n",
    "# Extrinsic validation evaluates a variant using external data, such as ensuring that the reference nucleotide in the variant matches that implied by the reference sequence and position.\n",
    "import hgvs.validator\n",
    "# Normalization is always 3’ with respect to the reference sequence. \n",
    "import hgvs.normalizer\n",
    "from hgvs.exceptions import HGVSError\n",
    "\n",
    "\n",
    "hp = hgvs.parser.Parser()\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "hn = hgvs.normalizer.Normalizer(hdp)\n",
    "vr = hgvs.validator.Validator(hdp=hdp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "hgvs.exceptions.HGVSParseError('NM_004343.4:c.1154_1155ins5: char 26: Syntax error')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def validate_hgvs_variants(expression):\n",
    "    try:\n",
    "        parsed_variant = hp.parse_hgvs_variant(expression)\n",
    "        vr.validate(parsed_variant)\n",
    "        return expression\n",
    "    except HGVSError as e:\n",
    "        return e\n",
    "validate_hgvs_variants('NM_004343.4:c.1154_1155ins5')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "excel_file = '../data/LabHgvsExpression.xlsx'\n",
    "\n",
    "with pd.ExcelFile(excel_file) as hgvs_path:\n",
    "    input_data = pd.read_excel(hgvs_path)\n",
    "\n",
    "input_data['oringal_hgvs_expression_example'] = input_data['oringal_hgvs_expression_example'].str.strip()\n",
    "input_data['edited_hgvs_expression_expample'] = input_data['edited_hgvs_expression_expample'].str.strip()\n",
    "\n",
    "data = pd.DataFrame(input_data)\n",
    "\n",
    "\n",
    "\n",
    "#NOTE: Boolean Values where obtained from: https://mutalyzer.nl/\n",
    "# Description: The Normalizer takes a variant description as input and checks whether it is correct.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_expression = data['oringal_hgvs_expression_example']\n",
    "edited_expression = data['edited_hgvs_expression_expample']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_hgvs_variants(hgvs_list):\n",
    "    # boolean_checker = []\n",
    "    error_messages = []\n",
    "\n",
    "    for hgvs in hgvs_list:\n",
    "        try:\n",
    "            parsed_variant = hp.parse_hgvs_variant(hgvs)\n",
    "            vr.validate(parsed_variant)\n",
    "            # boolean_checker.append()\n",
    "            error_messages.append('Passed')\n",
    "        except HGVSError as e:\n",
    "            # boolean_checker.append(False)\n",
    "            error_messages.append(e) \n",
    "    \n",
    "    return boolean_checker, error_messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "original_boolean_checker, original_error_messages = validate_hgvs_variants(original_expression)\n",
    "edited_boolean_checker, edited_error_messages = validate_hgvs_variants(edited_expression)\n",
    "\n",
    "hgvs_results = {\n",
    "    # 'biocommons_hgvs_original_hgvs_checker':original_boolean_checker,\n",
    "    'biocommons_hgvs_original_hgvs_error_messages':original_error_messages,\n",
    "    # 'biocommons_hgvs_edited_hgvs_checker':edited_boolean_checker,\n",
    "    'biocommons_hgvs_edited_hgvs_error_messages':edited_error_messages\n",
    "}\n",
    "\n",
    "biocommons_hgvs_test = pd.DataFrame(hgvs_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "biocommons_hgvs_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = pd.concat([data, biocommons_hgvs_test], axis=1, ignore_index=False)\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv('../data/results.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGVS</th>\n",
       "      <th>biocommons_validator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>c.80006G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>variant in the TTN(NM_001267550.2) gene</td>\n",
       "      <td>TTN(NM_001267550.2): char 6: expected one of '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A (p.Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2(TTN):c.80006G&gt;A : char 30: expe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A p.Ser26669Asn</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NM_001267550.2 p.Ser26669Asn</td>\n",
       "      <td>Accession (NM_001267550.2) is not compatible w...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C (p.Cys1913Arg)</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000059.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c.5737T&gt;C (p.C1913R) in BRCA2</td>\n",
       "      <td>c.5737T&gt;C : char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BRCA2:c.5737T&gt;C p.C1913R</td>\n",
       "      <td>BRCA2:p.C1913R : char 15: expected EOF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BRCA2 c.5737T&gt;C</td>\n",
       "      <td>Accession (BRCA2) is not known to be compatibl...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>BRCA2:C1913R</td>\n",
       "      <td>BRCA2:C1913R: char 6: expected one of 'c', 'g'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C (p.Glu74Lys)</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>p.E74K (Glu74Lys) c.220G&gt;C (g.29175102)</td>\n",
       "      <td>c.220G&gt;C: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Chr18:g.29175102</td>\n",
       "      <td>NC_000008.11:g.29175102: char 24: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TTR E74K</td>\n",
       "      <td>TTR:E74K : char 4: expected one of 'c', 'g', '...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TTR c.220G&gt;C</td>\n",
       "      <td>Accession (TTR) is not known to be compatible ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A (p.Val142Ile)</td>\n",
       "      <td>No transcript definition for (tx_ac=NM_000371.4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>p.V142I g.29178618</td>\n",
       "      <td>p.V142I: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>c.424G&gt;A p.Val142Ile</td>\n",
       "      <td>c.424G&gt;A: char 1: expected a letter or digit</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>TTR g.29178618</td>\n",
       "      <td>TTR:g.29178618: char 15: end of input</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GRCh37 Chr18:g.29178618</td>\n",
       "      <td>NC_000018.9:g.29178618: char 23: end of input</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              HGVS  \\\n",
       "0        NM_001267550.2:c.80006G>A (p.Ser26669Asn)   \n",
       "1                       c.80006G>A (p.Ser26669Asn)   \n",
       "2          variant in the TTN(NM_001267550.2) gene   \n",
       "3   NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn)   \n",
       "4          NM_001267550.2:c.80006G>A p.Ser26669Asn   \n",
       "5                     NM_001267550.2 p.Ser26669Asn   \n",
       "6             NM_000059.4:c.5737T>C (p.Cys1913Arg)   \n",
       "7                    c.5737T>C (p.C1913R) in BRCA2   \n",
       "8                         BRCA2:c.5737T>C p.C1913R   \n",
       "9                                  BRCA2 c.5737T>C   \n",
       "10                                    BRCA2:C1913R   \n",
       "11               NM_000371.4:c.220G>C (p.Glu74Lys)   \n",
       "12         p.E74K (Glu74Lys) c.220G>C (g.29175102)   \n",
       "13                                Chr18:g.29175102   \n",
       "14                                        TTR E74K   \n",
       "15                                    TTR c.220G>C   \n",
       "16              NM_000371.4:c.424G>A (p.Val142Ile)   \n",
       "17                              p.V142I g.29178618   \n",
       "18                            c.424G>A p.Val142Ile   \n",
       "19                                  TTR g.29178618   \n",
       "20                         GRCh37 Chr18:g.29178618   \n",
       "\n",
       "                                 biocommons_validator  \n",
       "0                                                True  \n",
       "1      c.80006G>A: char 1: expected a letter or digit  \n",
       "2   TTN(NM_001267550.2): char 6: expected one of '...  \n",
       "3   NM_001267550.2(TTN):c.80006G>A : char 30: expe...  \n",
       "4                                                True  \n",
       "5   Accession (NM_001267550.2) is not compatible w...  \n",
       "6    No transcript definition for (tx_ac=NM_000059.4)  \n",
       "7      c.5737T>C : char 1: expected a letter or digit  \n",
       "8              BRCA2:p.C1913R : char 15: expected EOF  \n",
       "9   Accession (BRCA2) is not known to be compatibl...  \n",
       "10  BRCA2:C1913R: char 6: expected one of 'c', 'g'...  \n",
       "11   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "12       c.220G>C: char 1: expected a letter or digit  \n",
       "13     NC_000008.11:g.29175102: char 24: end of input  \n",
       "14  TTR:E74K : char 4: expected one of 'c', 'g', '...  \n",
       "15  Accession (TTR) is not known to be compatible ...  \n",
       "16   No transcript definition for (tx_ac=NM_000371.4)  \n",
       "17        p.V142I: char 1: expected a letter or digit  \n",
       "18       c.424G>A: char 1: expected a letter or digit  \n",
       "19              TTR:g.29178618: char 15: end of input  \n",
       "20      NC_000018.9:g.29178618: char 23: end of input  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import hgvs.parser\n",
    "import hgvs.dataproviders.uta\n",
    "import hgvs.validator\n",
    "from hgvs.exceptions import HGVSError\n",
    "\n",
    "hp = hgvs.parser.Parser()\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "# TODO:open different type of files. If the file format doesn't work then throw an error.\n",
    "excel_file = '../data/LabHgvsExpression.xlsx'\n",
    "with pd.ExcelFile(excel_file) as hgvs_path:\n",
    "    input_data = pd.read_excel(hgvs_path)\n",
    "\n",
    "\n",
    "def validate_hgvs_variants(hgvs_list):\n",
    "    error_messages = []\n",
    "\n",
    "    for hgvs in hgvs_list:\n",
    "        try:\n",
    "            parsed_variant = hp.parse_hgvs_variant(hgvs)\n",
    "            vr.validate(parsed_variant)\n",
    "            error_messages.append(True)\n",
    "        except HGVSError as e:\n",
    "            error_messages.append(e) \n",
    "    \n",
    "    return error_messages\n",
    "\n",
    "#TODO:user inputes the column name of there hgvs expression\n",
    "input_data['oringal_hgvs_expression_example'] = input_data['oringal_hgvs_expression_example'].str.strip()\n",
    "\n",
    "data = pd.DataFrame(input_data)\n",
    "hgvs_expression = data['edited_hgvs_expression_expample']\n",
    "\n",
    "original_error_messages = validate_hgvs_variants(hgvs_expression)\n",
    "\n",
    "#TODO: create output file.\n",
    "hgvs_results = {'HGVS':data['oringal_hgvs_expression_example'],\n",
    "                'biocommons_validator':original_error_messages}\n",
    "biocommons_hgvs_test = pd.DataFrame(hgvs_results)\n",
    "biocommons_hgvs_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hgvs.validator\n",
    "import hgvs.exceptions\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "try:\n",
    "    vr.validate( hp.parse_hgvs_variant('NM_000371.4:c.220G>C') )\n",
    "except hgvs.exceptions.HGVSError as e:\n",
    "    print(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hgvs_list = ['NC_000017.11:g.43091687delC', 'NC_000007.13.g.21726874G>A']\n",
    "\n",
    "try:\n",
    "    parser = hgvs.parser.Parser()\n",
    "    validator = hgvs.validator.IntrinsicValidator()\n",
    "    \n",
    "    for hgvs in hgvs_list:\n",
    "        variant = parser.parse_hgvs_variant(hgvs)\n",
    "        validator.validate(variant)\n",
    "except Exception as e:\n",
    "    print(e)\n",
    "\n",
    "import hgvs.validator\n",
    "import hgvs.exceptions\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "try:\n",
    "    vr.validate( hp.parse_hgvs_variant('NM_001267550.2(TTN):c.80006G>A') )\n",
    "except hgvs.exceptions.HGVSError as e:\n",
    "    print(e)\n",
    "\n",
    "\n",
    "from hgvs.exceptions import HGVSError\n",
    "import hgvs.parser\n",
    "import hgvs.validator\n",
    "\n",
    "hgvs_list = ('NC_000017.11:g.43091687delC', 'NC_000007.13.g.21726874G>A')\n",
    "\n",
    "# Create a validator instance\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "# Create a parser instance\n",
    "hp = hgvs.parser.Parser()\n",
    "\n",
    "try:\n",
    "    for hgvs in hgvs_list:\n",
    "        vr.validate(hp.parse_hgvs_variant(hgvs))\n",
    "        print('expression passed: {}'.format(hgvs))\n",
    "except HGVSError as e:\n",
    "    print(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import hgvs.parser\n",
    "import hgvs.dataproviders.uta\n",
    "import hgvs.validator\n",
    "from hgvs.exceptions import HGVSError\n",
    "\n",
    "# import sys\n",
    "# sys.path.append('..')\n",
    "\n",
    "hp = hgvs.parser.Parser()\n",
    "hdp = hgvs.dataproviders.uta.connect()\n",
    "vr = hgvs.validator.Validator(hdp=hdp)\n",
    "\n",
    "excel_file = '/Users/M278428/Documents/rf_lab_projects/DraftCoreDataModel/data/finaltestdata.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_data = pd.read_excel(excel_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hgvs_expression</th>\n",
       "      <th>human_curated</th>\n",
       "      <th>Failure Reason</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A, p.(Ser26669Asn)</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C, p.(Cys1913Arg)</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C, p.(Glu74Lys)</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A, p.(Val142Ile)</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000371.4:c.148G&gt;A, p.(Val50Met)</td>\n",
       "      <td>pass</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...</td>\n",
       "      <td>fail</td>\n",
       "      <td>no refseq, incorrect format for mosaic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>CALR c.1154insTTGTC (mosaic)</td>\n",
       "      <td>fail</td>\n",
       "      <td>no refseq, incorrect allelic formatting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>EZH2: c.2196A&gt;G (homozygous)</td>\n",
       "      <td>fail</td>\n",
       "      <td>mismatch between refseq and variant type (mito...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NC_012920.1:c.1555A&gt;G</td>\n",
       "      <td>fail</td>\n",
       "      <td>incorrect punctuation (should use brackets for...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>NC_000023:g.147912051CGG(33)</td>\n",
       "      <td>fail</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       hgvs_expression human_curated  \\\n",
       "0           NM_001267550.2:c.80006G>A, p.(Ser26669Asn)          pass   \n",
       "1                NM_000059.4:c.5737T>C, p.(Cys1913Arg)          pass   \n",
       "2                   NM_000371.4:c.220G>C, p.(Glu74Lys)          pass   \n",
       "3                  NM_000371.4:c.424G>A, p.(Val142Ile)          pass   \n",
       "4                   NM_000371.4:c.148G>A, p.(Val50Met)          pass   \n",
       "..                                                 ...           ...   \n",
       "96   NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...          fail   \n",
       "97                        CALR c.1154insTTGTC (mosaic)          fail   \n",
       "98                        EZH2: c.2196A>G (homozygous)          fail   \n",
       "99                               NC_012920.1:c.1555A>G          fail   \n",
       "100                       NC_000023:g.147912051CGG(33)          fail   \n",
       "\n",
       "                                        Failure Reason  \n",
       "0                                                  NaN  \n",
       "1                                                  NaN  \n",
       "2                                                  NaN  \n",
       "3                                                  NaN  \n",
       "4                                                  NaN  \n",
       "..                                                 ...  \n",
       "96              no refseq, incorrect format for mosaic  \n",
       "97             no refseq, incorrect allelic formatting  \n",
       "98   mismatch between refseq and variant type (mito...  \n",
       "99   incorrect punctuation (should use brackets for...  \n",
       "100                                                NaN  \n",
       "\n",
       "[101 rows x 3 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_data['hgvs_expression'] = input_data['hgvs_expression'].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0             NM_001267550.2:c.80006G>A, p.(Ser26669Asn)\n",
       "1                  NM_000059.4:c.5737T>C, p.(Cys1913Arg)\n",
       "2                     NM_000371.4:c.220G>C, p.(Glu74Lys)\n",
       "3                    NM_000371.4:c.424G>A, p.(Val142Ile)\n",
       "4                     NM_000371.4:c.148G>A, p.(Val50Met)\n",
       "                             ...                        \n",
       "96     NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...\n",
       "97                          CALR c.1154insTTGTC (mosaic)\n",
       "98                          EZH2: c.2196A>G (homozygous)\n",
       "99                                 NC_012920.1:c.1555A>G\n",
       "100                         NC_000023:g.147912051CGG(33)\n",
       "Name: hgvs_expression, Length: 101, dtype: object"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgvsExamples = input_data['hgvs_expression']\n",
    "hgvsExamples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_hgvs_variants(hgvs_list):\n",
    "    error_messages = []\n",
    "\n",
    "    for hgvs in hgvs_list:\n",
    "        try:\n",
    "            parsed_variant = hp.parse_hgvs_variant(hgvs)\n",
    "            vr.validate(parsed_variant)\n",
    "            error_messages.append(True)\n",
    "        except HGVSError as e:\n",
    "            error_messages.append(e) \n",
    "    \n",
    "    return error_messages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[hgvs.exceptions.HGVSParseError('NM_001267550.2:c.80006G>A, p.(Ser26669Asn): char 25: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000059.4:c.5737T>C, p.(Cys1913Arg): char 21: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000371.4:c.220G>C, p.(Glu74Lys): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000371.4:c.424G>A, p.(Val142Ile): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000371.4:c.148G>A, p.(Val50Met): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000097.7:c.814A>C, p.(Asn272His): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_007194.4:c.470T>C, p.(Ile157Thr): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_001122764.3:c.767C>G, p.(Pro256Arg): char 23: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_004333.6:c.1799T>A, p.(Val600Glu): char 21: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000077.5:c.335G>C, p.(Arg112Pro): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_000546.6:c.[856G>A];[833C>T];(46delC), NP_000537.3:p.[Glu286Lys];[Pro278Leu];(Gln16fs): char 14: expected one of '(', '*', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_000039.3:c.296T>C, p.(Leu99Pro): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_000041.4:c.388T>C, p.(Cys130Arg): char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_001256850.1:c.1141G>A, p.(Gly381Ser): char 24: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_015338.6:c.1934=/dupG, p.(Gly646=/Trpfs*12): char 19: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_001127208.3:c.(3940=/delG);[4589_4618del30],: char 22: expected one of ')', '_', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NP_001120680.1:p.((Asp1314=/Metfs*49));[(Pro1530_Gln1539del)]: char 26: expected one of ')', '_', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_004343.4:c.1154_1155=/insTTGTC , NP_004334.1:p.(Lys385=/Asnfs*47): char 24: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_004456.5:c.[2196-2A>G];[2196-2A>G]: char 14: expected one of '(', '*', or a digit\"),\n",
       " True,\n",
       " hgvs.exceptions.HGVSParseError(\"NC_000023.11:g.147912051CGG[33]: char 27: expected the character '='\"),\n",
       " hgvs.exceptions.HGVSParseError('c.80006G>A (p.Ser26669Asn) variant in the TTN(NM_001267550.2) gene: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('c.5737T>C (p.C1913R) in BRCA2: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('p.E74K (Glu74Lys) c.220G>C (g.29175102): char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('p.V142I g.29178618: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('p.V50M (p.Val50Met): char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('c.814A>C (p.N272H) alteration is a polymorphism of the coproporphyrinogen oxidase gene: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('p.Ile157Thr: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError(\"PPOX c.767C>G p.P256R: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"BRAF DNA change: c.1799T>A Amino Acid change: p.V600E: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"CDKN2A c.335G>C p.R112P: char 6: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_000546.6:c.[856G>A];[833C>T];(46delC): char 14: expected one of '(', '*', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_000039.3:c.296T/C: char 18: expected one of '=', or '>'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"APOE e4: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TTN c.1141G>A p.Gly381Ser: char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_015338.6:c.1934dupG (25%): char 23: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_001127208.3:c.(3940=/delG);[4589_4618del30]: char 22: expected one of ')', '_', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_004343.4:c.1154_1155=/insTTGTC: char 24: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_004456.5:c.2196-2A>G (homozygous): char 23: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"MT:A1555AG: char 3: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"FMR1:-129CGG(10>33): char 5: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_001267550.2(TTN):c.80006G>A (p.Ser26669Asn): char 30: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('BRCA2:c.5737T>C p.C1913R: char 15: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('Chr18:g.29175102: char 17: end of input'),\n",
       " hgvs.exceptions.HGVSParseError('c.424G>A p.Val142Ile: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('c.148G>A (g.29172937): char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('c.814A>C (p.N272H): char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError(\"CHEK2 p.Ile157Thr: char 5: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"PPOX (Pro256Arg) c.767C>G: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('BRAF:c.1799T>A (Val600Glu): char 14: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"CDKN2A G335C (R112P): char 6: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_000546.6:c.856G>A, c.833C>T, c.46delC: char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_000039.3 T296C (L99P): char 11: expected one of '(', ':', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"APOE p.Cys112Arg: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_001256850 c1141G>A: char 12: expected one of '(', '.', ':', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_015338.6 c.1934/dupG: char 11: expected one of '(', ':', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_001127208.3:c.4589_4618del30, c.3940delG: char 32: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError('NM_004343.4:c.1154_1155ins5: char 26: Syntax error'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_004456.6 c.2196-2A>G/G: char 11: expected one of '(', ':', or a digit\"),\n",
       " hgvs.exceptions.HGVSInvalidVariantError('Accession (MTRNR1) is not known to be compatible with variant type c'),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_002024.6:c.-129CGG[33]: char 21: expected the character '='\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_001267550.2:c.80006G>A p.Ser26669Asn: char 25: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"BRCA2 c.5737T>C: char 5: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TTR E74K: char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TTR g.29178618: char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('c.148G>A (p.Val50Met): char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('c.814A>C (p.Asn272His) coproporphyrinogen oxidase: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError(\"CHEKZ p.Ile157Thr: char 5: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"PPOX (Pro256Arg) (g.161138933): char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"BRAF c.1799T>A p.V600E: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"CDKN2A 335G/C: char 6: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_000546.6:c.856G>A + 833C>T + 46delC: char 20: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"APO1A:296T>C (Leu99Pro): char 6: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"APOE c.388T>C (p.Cys130Arg): char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_001256850.1:1141G>A: char 15: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_015338.6:1934=/dupG: char 12: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TET2:3940delG;4589_4618del30: char 5: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_004343.4:c.1154_1155ins5 p.(Lys385fs): char 26: Syntax error'),\n",
       " hgvs.exceptions.HGVSParseError(\"EZH2 IVS19-2A>G: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NC_012920:m1555A>G: char 11: expected the character '.'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_002024.6:c.[-158_-129dup30; -129insCGGCGGCGG]: char 14: expected one of '(', '*', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_001267550.2 p.Ser26669Asn: char 14: expected one of '(', ':', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"BRCA2:C1913R: char 6: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TTR c.220G>C: char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"GRCh37 Chr18:g.29178618: char 6: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"TTR (p.Val50Met): char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError('c.814A>C (p.N272H) CPOX: char 1: expected a letter or digit'),\n",
       " hgvs.exceptions.HGVSParseError('I157T: char 6: end of input'),\n",
       " hgvs.exceptions.HGVSParseError(\"PPOX c.767C>C/G: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"BRAF c.1799A V600E: char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"CDKN2A p.Arg112Pro: char 6: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_000546.6:c.[856G>A][833C>T](46delC): char 14: expected one of '(', '*', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NM_000039.3:c.296C: char 19: end of input'),\n",
       " hgvs.exceptions.HGVSDataNotAvailableError('No transcript definition for (tx_ac=NM_000041.4)'),\n",
       " hgvs.exceptions.HGVSParseError(\"TTN p.381G/S: char 3: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"NM_015338.6:c.WT(0.75)/1934dupG(0.25): char 14: expected one of '(', '*', or a digit\"),\n",
       " hgvs.exceptions.HGVSParseError('NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1539del]: char 33: expected EOF'),\n",
       " hgvs.exceptions.HGVSParseError(\"CALR c.1154insTTGTC (mosaic): char 4: expected one of '(', '.', ':', '_', or a letter or digit\"),\n",
       " hgvs.exceptions.HGVSParseError(\"EZH2: c.2196A>G (homozygous): char 5: expected one of 'c', 'g', 'm', 'n', 'p', or 'r'\"),\n",
       " hgvs.exceptions.HGVSInvalidVariantError('Accession (NC_012920.1) is not compatible with variant type c'),\n",
       " hgvs.exceptions.HGVSParseError(\"NC_000023:g.147912051CGG(33): char 24: expected the character '='\")]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = validate_hgvs_variants(hgvsExamples)\n",
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HGVS</th>\n",
       "      <th>Validator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A, p.(Ser26669Asn)</td>\n",
       "      <td>NM_001267550.2:c.80006G&gt;A, p.(Ser26669Asn): ch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_000059.4:c.5737T&gt;C, p.(Cys1913Arg)</td>\n",
       "      <td>NM_000059.4:c.5737T&gt;C, p.(Cys1913Arg): char 21...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_000371.4:c.220G&gt;C, p.(Glu74Lys)</td>\n",
       "      <td>NM_000371.4:c.220G&gt;C, p.(Glu74Lys): char 20: e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_000371.4:c.424G&gt;A, p.(Val142Ile)</td>\n",
       "      <td>NM_000371.4:c.424G&gt;A, p.(Val142Ile): char 20: ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_000371.4:c.148G&gt;A, p.(Val50Met)</td>\n",
       "      <td>NM_000371.4:c.148G&gt;A, p.(Val50Met): char 20: e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...</td>\n",
       "      <td>NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>CALR c.1154insTTGTC (mosaic)</td>\n",
       "      <td>CALR c.1154insTTGTC (mosaic): char 4: expected...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>EZH2: c.2196A&gt;G (homozygous)</td>\n",
       "      <td>EZH2: c.2196A&gt;G (homozygous): char 5: expected...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>NC_012920.1:c.1555A&gt;G</td>\n",
       "      <td>Accession (NC_012920.1) is not compatible with...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>NC_000023:g.147912051CGG(33)</td>\n",
       "      <td>NC_000023:g.147912051CGG(33): char 24: expecte...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  HGVS  \\\n",
       "0           NM_001267550.2:c.80006G>A, p.(Ser26669Asn)   \n",
       "1                NM_000059.4:c.5737T>C, p.(Cys1913Arg)   \n",
       "2                   NM_000371.4:c.220G>C, p.(Glu74Lys)   \n",
       "3                  NM_000371.4:c.424G>A, p.(Val142Ile)   \n",
       "4                   NM_000371.4:c.148G>A, p.(Val50Met)   \n",
       "..                                                 ...   \n",
       "96   NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...   \n",
       "97                        CALR c.1154insTTGTC (mosaic)   \n",
       "98                        EZH2: c.2196A>G (homozygous)   \n",
       "99                               NC_012920.1:c.1555A>G   \n",
       "100                       NC_000023:g.147912051CGG(33)   \n",
       "\n",
       "                                             Validator  \n",
       "0    NM_001267550.2:c.80006G>A, p.(Ser26669Asn): ch...  \n",
       "1    NM_000059.4:c.5737T>C, p.(Cys1913Arg): char 21...  \n",
       "2    NM_000371.4:c.220G>C, p.(Glu74Lys): char 20: e...  \n",
       "3    NM_000371.4:c.424G>A, p.(Val142Ile): char 20: ...  \n",
       "4    NM_000371.4:c.148G>A, p.(Val50Met): char 20: e...  \n",
       "..                                                 ...  \n",
       "96   NP_001120680.1:p.Asp1314Metfs*49;[Pro1530_Gln1...  \n",
       "97   CALR c.1154insTTGTC (mosaic): char 4: expected...  \n",
       "98   EZH2: c.2196A>G (homozygous): char 5: expected...  \n",
       "99   Accession (NC_012920.1) is not compatible with...  \n",
       "100  NC_000023:g.147912051CGG(33): char 24: expecte...  \n",
       "\n",
       "[101 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hgvs_results = {\n",
    "    'HGVS': hgvsExamples,\n",
    "    'Validator': results\n",
    "}\n",
    "result = pd.DataFrame(hgvs_results) \n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_dev_vrs_installation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
